File: online_judges/kattis/classy.cpp
#include <bits/stdc++.h>

using namespace std;
struct P{
	string name;
	string clas;
};
bool cmp(P p1,P p2){
	if(p1.clas!=p2.clas) return p1.clas < p2.clas;
	return p1.name > p2.name;
}
string reverse(string s){
	string S = "";
	for(int i=s.size()-1;i>=0;i--){
		S+=s[i];
	}
	return S;
}
int main(){
//	freopen("i.in","r",stdin);
	int t,n,lmax,sz;
	string name,clas,s,str;
	cin >> t;
	while(t--){
		lmax = 0;
		cin >> n;
		vector<P> H(n);
		for(int i=0;i<n;i++){
			
			cin >> name >> clas >> str;
			name.replace(name.size()-1,1,"");
			H[i].name = name;
			s = "";
			int p = clas.find("-");
			while(p!=string::npos){
				clas.replace(p,1," ");
				p = clas.find("-");
			}
			istringstream iss(clas);
			do{
				string str;
				iss >> str;
				if(str=="upper") s+="3";
				if(str=="lower") s+="1";
				if(str=="middle") s+="2";
			}while(iss);
			lmax = max((int)s.size(),lmax);
			H[i].clas = s;
		}

		for(int i=0;i<n;i++){
			s = H[i].clas;
			s = reverse(s);
			sz = (int)s.size();
			for(int j=1;j<=lmax-sz;j++){
				s += "2";
			}
			H[i].clas = s;
		}
		sort(H.begin(),H.end(),cmp);
		for(int i=n-1;i>=0;i--){
//			cout<<H[i].name<<" "<<H[i].clas<<endl;
			cout<<H[i].name<<endl;
//			cout<<H[i].clas<<endl;
		}
		cout<<"==============================\n";
	}
	return 0;
}

File: online_judges/kattis/classy.java

import java.io.*;
import java.util.*;


public class Solution {
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static int mod = (1 << 31) - 1;
    static List<Integer>[] graph;

    static char[][] grid;


    public static void main(String[] args){
        InputStream is;
        try {
            is = new FileInputStream(".\\src\\in.txt");
        } catch (FileNotFoundException e) {
            is = System.in;
        }
        InputReader in = new InputReader(is);
        PrintWriter out = new PrintWriter(System.out);

        int t = in.nextInt();
        StringBuilder sb = new StringBuilder();
        while (t-- > 0) {
            int n = in.nextInt();
            P[] H = new P[n];
            int lmax = 0;
            for (int i=0;i<n;i++) {
                String name = in.next();
                name = name.substring(0, name.length()-1);
                String clas = in.next();
                String str = in.next();
                String s = "";
                String[] parts = clas.split("-");
                for (String part: parts) {
                    if(part.equals("upper")) s += "3";
                    else if (part.equals("middle")) s += "2";
                    else s += "1";
                }
                lmax = Math.max(lmax, s.length());
                H[i] = new P(name, s);
            }

            for(int i=0;i<n;i++){
                String s = H[i].clas;
                s = reverse(s);
                int sz = s.length();
                for(int j=1;j<=lmax-sz;j++){
                    s += "2";
                }
                H[i].clas = s;
            }
            Arrays.sort(H);
            for(int i=n-1;i>=0;i--) sb.append(H[i].name + "\n");
            sb.append("==============================\n");
        }

        out.print(sb);
        out.close();
    }

    static class P implements Comparable{
        public String name, clas;

        public P(String name, String clas) {
            this.name = name;
            this.clas = clas;
        }

        @Override
        public int compareTo(Object o) {
            P p  = (P)o;
            if (clas.equals(p.clas)) {
                return name.compareTo(p.name) * -1;
            }
            return clas.compareTo(p.clas);
        }
    }
    static String reverse(String s){
        String S = "";
        for(int i=s.length()-1;i>=0;i--){
            S+=s.charAt(i);
        }
        return S;
    }

    static class InputReader {
        public BufferedReader reader;
        public StringTokenizer tokenizer;

        public InputReader(InputStream stream) {
            reader = new BufferedReader(new InputStreamReader(stream), 32768);
            tokenizer = null;
        }

        public String next() {
            while (tokenizer == null || !tokenizer.hasMoreTokens()) {
                try {
                    tokenizer = new StringTokenizer(reader.readLine());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }
            return tokenizer.nextToken();
        }

        public int nextInt() {
            return Integer.parseInt(next());
        }
        public long nextLong() {
            return Long.parseLong(next());
        }
        public double nextDouble() {return Double.parseDouble(next());}
    }
}